### PR TITLE
mm checks and correct int dtypes if A is on the gpu

### PR DESCRIPTION
**Pull request by coquelin77**
## Description

fixes bug by casting tensors to floats before mm operation

Issue/s resolved: #657 

## Changes proposed:
- for `matmul`: cast tensors to floats if the devices are GPUs and the dtypes are ints

## Type of change
- Bug fix (non-breaking change which fixes an issue)

## Due Diligence

- [x] All split configurations tested
- [x] Multiple dtypes tested in relevant functions
- [x] Documentation updated (if needed)
- [x] Updated changelog.md under the title "Pending Additions"

#### Does this change modify the behaviour of other functions? If so, which?
no

### CODE REVIEW

**Comment by mtar** [comment]
GPU cluster tests are currently disabled on this Pull Request.

**Comment by mtar** [comment]
test this please

**Comment by bhagemeier** [comment]
It works, but the results are now dependent on whether you compute on the GPU or CPU. Do we have other places where we do similar stuff? The actual difference at first sight is the difference in the data type returned (int64 vs. float64). Further down the road, we're losing 11 bits of precision for very large numbers. This may be a corner case, but I wanted to point it out before approving this.

```
>>> a = ht.array([[0, 1], [2, 3]])
>>> b = ht.array([[4, 5], [6, 7]])
>>> ht.dot(a, b)
DNDarray([[ 6,  7],
          [26, 31]], dtype=ht.int64, device=cpu:0, split=None)
>>> a.gpu()
DNDarray([[0, 1],
          [2, 3]], dtype=ht.int64, device=gpu:0, split=None)
>>> b.gpu()
DNDarray([[4, 5],
          [6, 7]], dtype=ht.int64, device=gpu:0, split=None)
>>> ht.dot(a, b)
DNDarray([[ 6.,  7.],
          [26., 31.]], dtype=ht.float64, device=gpu:0, split=None)
```

**Comment by coquelin77** [comment]
as far as i know, this is the only place where we do a specific type change for GPUs. however, the function required for this operation, namely `addmm_cuda` is not implemented so we are forced to work around it. even if we were to do this:
```python
r = a_block @ b_block
c[c_start0 : c_start0 + mB, c_start1 : c_start1 + nB] += r
```
it still fails with the same error, unfortunately I do not see a way around this. we can loop back to it when this function is implemented in torch.

**Comment by mtar** [comment]
The dtype of the returned array changes. As a user, I would expect it to be the same as the input.

**Comment by coquelin77** [comment]
again, the operation which we need is **not implemented** in torch. this **cannot** be done a different way in this specific scenario. It can throw a warning but that seems excessive.

**Comment by mtar** [comment]
You can change the type afterwards

**Comment by coquelin77** [comment]
the precision is already lost even if the type is changed back. we can cast back at the end but it does require yet another check in all 8 return locations

**Comment by Markus-Goetz** [comment]
Even if we do loose precision, we should be consistent on the types. For GPU, cast the output tensor back to int64. The case that you actually do int64 mm on GPU is relatively seldom to being with anyway.

**Comment by mtar** [comment]
rerun tests

**Comment by coquelin77** [comment]
this now casts back to the initial promoted dtype for ints on GPUs

File: CHANGELOG.md [hunk 1]
```diff
@@ -58,10 +59,6 @@
 - [#652](https://github.com/helmholtz-analytics/heat/pull/652) Feature: benchmark scripts and jobscript generation
 - [#653](https://github.com/helmholtz-analytics/heat/pull/653) Printing above threshold gathers the data without a buffer now
 - [#653](https://github.com/helmholtz-analytics/heat/pull/653) Bugfixes: Update unittests argmax & argmin + force index order in mpi_argmax & mpi_argmin. Add device parameter for tensor creation in dndarray.get_halo().
-- [#659](https://github.com/helmholtz-analytics/heat/pull/659) New feature: distributed `random.permutation` + `random.randperm`
-- [#662](https://github.com/helmholtz-analytics/heat/pull/662) Bugfixes: `minimum()` and `maximum()` split semantics, scalar input, different input dtype
-- [#664](https://github.com/helmholtz-analytics/heat/pull/664) New feature / enhancement: distributed `random.random_sample`, `random.random`, `random.sample`, `random.ranf`, `random.random_integer`
-- [#666](https://github.com/helmholtz-analytics/heat/pull/666) New feature: distributed prepend/append for `diff()`.
```
Comment (bhagemeier) [inline CHANGELOG.md:64]: These lines should not have been removed from the CHANGELOG. I'll fix that.

Comment (coquelin77) [reply CHANGELOG.md:64]: oh sorry! i thought they were duplicates. my bad

**Review by bhagemeier — DISMISSED** [review]
Looks good now from my point of view.

**Review by bhagemeier — APPROVED** [review]
Test have been added. I have a feeling that codecov is having trouble just now. Therefore, I vote for ignoring the minute (.01%) decrease in coverage.

**Pull request merged**